### ISSUE TITLE
the @tf.function annotation weird performance

### ISSUE THREAD

**Issue by gao8954** [tensorflow/tensorflow#51898]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- Ubuntu 18.04.5
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):
- binary
- TensorFlow version (use command below):
- Python version:
- Python 3.6.9
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

tf version 2.6.0

**Describe the current behavior**

tf version 2.6

when the model call function has no @tf.function annotation like this:
![image](https://user-images.githubusercontent.com/7404433/132669008-2043ccca-c588-4809-bbd9-e8255d0b3342.png)

train the model the loss and auc log is like this:

![image](https://user-images.githubusercontent.com/7404433/132669307-b105084e-ec5a-4084-bf21-fe87a65feac2.png)

as you can see from the picture,at epoch start the loss value is high, after epoch end the loss value is low,and auc >0.5
![image](https://user-images.githubusercontent.com/7404433/132669640-23a22f16-ec0c-4920-8780-204eb859edc2.png)

**All the code remains unchanged, just add @tf.function annotation like this:** 
![image](https://user-images.githubusercontent.com/7404433/132669988-e8725067-e4e0-40cd-b701-ae38c0b555d2.png)

train the model the loss and auc log is like this:
![image](https://user-images.githubusercontent.com/7404433/132670270-689f4e48-132a-4462-9c4d-ba891d91e125.png)

as you can see from the picture,at epoch start the loss value is different ,Its value is much lower，and after epoch end the loss value **is optimizer much lower, but the auc is always 0.5**

![image](https://user-images.githubusercontent.com/7404433/132670962-4b2c520b-6252-472a-a3ca-374dbb312c0d.png)

it is too confusing!!!

**Describe the expected behavior**

I think their results should be exactly the same，

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no):
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

The code needed to reproduce this problem：

Just modify the @tf.function annotation (add or remove) reproduce this problem

![image](https://user-images.githubusercontent.com/7404433/132793559-66cc46cc-1544-4330-9089-aef2d80ab6bb.png)

https://colab.research.google.com/drive/1i6PnUrkvNszW03wnS6ZyQBsrI1HYnTpq?usp=sharing

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by mohantym** [comment]
Ok. @gao8954,
But Could you please fill the issue template ?It helps us analyse the issue [tf version, steps followed before you ran into this error or stand alone code/colab gist to reproduce the issue faced].Thanks

**Comment by gao8954** [comment]
> Ok. @gao8954,
> But Could you please fill the issue template ?It helps us analyse the issue [tf version, steps followed before you ran into this error or stand alone code/colab gist to reproduce the issue faced].Thanks

Ok,has been modified,and colab link provided

**Comment by mohantym** [comment]
Ok @gao8954 ,Could you please provide link to csv file too?

**Comment by gao8954** [comment]
> Ok @gao8954 ,Could you please provide link to csv file too?

i have upload the csv file to the colab,but it always delete automatically,when exit the runtime

**Comment by gao8954** [comment]
[tmk_sample_50000.csv](https://github.com/tensorflow/tensorflow/files/7183704/tmk_sample_50000.csv)

i have uploaded again,and upload a backup here,just in case

**Comment by mohantym** [comment]
Hi @gao8954, I was getting different error  in TF 2.6 though , Attaching [gist ] (https://colab.research.google.com/gist/mohantym/7e304b7b98370a6fb18bca732383f800/untitled0.ipynb#scrollTo=W5VQnlisn0Y-)for reference. Can you try again after normalizing the data from CSV ( I saw some float point values among integer values) in input data.
